### PR TITLE
Run critical operations in transaction to showcase how to do that

### PR DESCRIPTION
**Pull request by jerben**

### CODE REVIEW

File: sihl_users/lib/service.ml [hunk 1]
```diff
@@ -209,20 +209,25 @@ module User = struct
...
+            in
+            Repository.User.update (Model.User.confirm user) connection)
       in
-      let user =
-        user |> Sihl_core.Fail.with_bad_request "invalid token provided"
-      in
-      Repository.User.update (Model.User.confirm user)
-      |> Sihl_core.Db.query_db_exn request
+      result
+      |> Sihl_core.Fail.with_bad_request "invalid token provided"
```
Comment (aronerben) [inline sihl_users/lib/service.ml:229]: @jerben Why is this line necessary? This was added in this diff

File: sihl_users/lib/service.ml [hunk 2]
```diff
@@ -209,20 +209,25 @@ module User = struct
...
+              |> Sihl_core.Db.query_db_exn request
+            in
+            let* user =
+              Repository.User.get ~id:token.user
+              |> Sihl_core.Db.query_db request
+            in
+            let user =
+              user |> Sihl_core.Fail.with_bad_request "invalid token provided"
+            in
+            Repository.User.update (Model.User.confirm user) connection)
```
Comment (aronerben) [inline sihl_users/lib/service.ml:226]: @jerben Out of curiosity: When are connections closed?

File: sihl_core/lib/db.ml [hunk 3]
```diff
@@ -105,8 +95,17 @@ let middleware app =
   let m = Rock.Middleware.create ~name:"database connection" ~filter in
   Opium.Std.middleware m app
 
-(* Execute a query on the database connection stored in the request
-   environment *)
+let query_db_with_trx request query =
+  let ( let* ) = Lwt.bind in
+  let connection = Request.env request |> Opium.Hmap.get key in
+  (fun connection ->
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl_core/lib/db.ml]: @aronerben Good catch, this is where I ended up following the types :) Fixed.

File: sihl_users/lib/service.ml (hunk 1 above)
Comment (jerben) [reply sihl_users/lib/service.ml]: @aronerben The new `Sihl_core.Db.query_db_with_trx` doesn't have an `_exn` version, I added it. Error handling is an open topic atm and not done very consistently.

File: sihl_users/lib/service.ml (hunk 2 above)
Comment (jerben) [reply sihl_users/lib/service.ml]: @aronerben Very good point! We need to stress test the whole thing to find leaking connections. All connections are retrieved from a connection pool so they should be managed by it. We don't drain it atm anywhere, the end of a Sihl application life cycle hasn't been implemented really yet.

Comment (jerben) [reply sihl_users/lib/service.ml]: When I close the connection on the MariaDB driver it segfaults even, on Postgres it works.

File: sihl_core/lib/db.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (aronerben) [reply sihl_core/lib/db.ml]: Thanks!

File: sihl_users/lib/service.ml (hunk 1 above)
Comment (aronerben) [reply sihl_users/lib/service.ml]: Ok!

File: sihl_users/lib/service.ml (hunk 2 above)
Comment (aronerben) [reply sihl_users/lib/service.ml]: So it's a Caqti problem?

Comment (jerben) [reply sihl_users/lib/service.ml]: @aronerben I don't know yet. I think it is wrong usage to release a connection while we retrieved it using `Pool.use (fun connection -> ...)`. The underlying bindings are directly wrapping the C drivers, maybe there is something wrong.
